File: backend/src/app/adapters/docx/contracts.py
"""Private contracts for the stateless DOCX processor."""

from __future__ import annotations

from enum import StrEnum

from app.domain.finding import DocxTextLocator
from app.sources.docx_text import DocxStoryKind, DocxTextBlock
from pydantic import BaseModel, Field, model_validator


class DocxProcessorOperation(StrEnum):
    HANDSHAKE = "handshake"
    INSPECT = "inspect"
    RENDER = "render"


class DocxProcessorErrorCode(StrEnum):
    PROTOCOL_ERROR = "protocol_error"
    PROCESSOR_FAILURE = "processor_failure"
    INVALID_DOCX = "invalid_docx"
    INVALID_TARGET = "invalid_target"
    OVERLAPPING_TARGETS = "overlapping_targets"
    INVALID_EFFECT = "invalid_effect"
    INVALID_REPLACEMENT = "invalid_replacement"
    UNSAFE_PATH = "unsafe_path"
    SOURCE_HASH_MISMATCH = "source_hash_mismatch"
    UNSUPPORTED_DOCX_FEATURE = "unsupported_docx_feature"


class ProcessorError(BaseModel):
    code: DocxProcessorErrorCode
    message: str
    feature: str | None = None

    model_config = {"extra": "forbid", "frozen": True}


class ProcessorResponse(BaseModel):
    request_id: str
    ok: bool
    payload: dict[str, object] | None
    error: ProcessorError | None
    working_set_bytes: int = Field(ge=0)

    model_config = {"extra": "forbid", "frozen": True}


class HandshakeResult(BaseModel):
    openxml_sdk_version: str
    operations: tuple[DocxProcessorOperation, ...]

    model_config = {"extra": "forbid", "frozen": True}


class SanitationSummary(BaseModel):
    comments_removed: int = Field(ge=0)
    properties_removed: int = Field(ge=0)

    model_config = {"extra": "forbid", "frozen": True}


class ProcessorImageOccurrence(BaseModel):
    """Processor-private structural picture evidence awaiting bitmap normalization."""

    id: str = Field(pattern=r"^pic_[0-9a-f]{64}$")
    ordinal: int = Field(ge=0)
    story_kind: DocxStoryKind
    part_uri: str = Field(min_length=1)
    media_type: str = Field(min_length=1)
    source_asset: str | None = Field(default=None, pattern=r"^[A-Za-z0-9._-]+$")
    crop_left: int = Field(ge=0, le=100000)
    crop_top: int = Field(ge=0, le=100000)
    crop_right: int = Field(ge=0, le=100000)
    crop_bottom: int = Field(ge=0, le=100000)
    flip_horizontal: bool
    flip_vertical: bool
    targetable: bool
    unsupported_reason: str | None = None

    @model_validator(mode="after")
    def validate_targetability(self) -> ProcessorImageOccurrence:
        if self.targetable != (self.source_asset is not None and self.unsupported_reason is None):
            raise ValueError("processor picture targetability is inconsistent")
        return self

    model_config = {"extra": "forbid", "frozen": True}


class DocxInspectResult(BaseModel):
    character_count: int = Field(ge=0)
    blocks: tuple[DocxTextBlock, ...]
    image_occurrences: tuple[ProcessorImageOccurrence, ...]
    document_sha256: str
    sanitation: SanitationSummary

    @model_validator(mode="after")
    def validate_inspection(self) -> DocxInspectResult:
        if tuple(block.ordinal for block in self.blocks) != tuple(range(len(self.blocks))):
            raise ValueError("DOCX blocks must have contiguous ordinals")
        if len({block.block_id for block in self.blocks}) != len(self.blocks):
            raise ValueError("DOCX block IDs must be unique")
        if self.character_count != sum(len(block.text) for block in self.blocks):
            raise ValueError("DOCX character count must match canonical blocks")
        if tuple(item.ordinal for item in self.image_occurrences) != tuple(
            range(len(self.image_occurrences))
        ):
            raise ValueError("DOCX picture occurrences must have contiguous ordinals")
        if len({item.id for item in self.image_occurrences}) != len(self.image_occurrences):
            raise ValueError("DOCX picture occurrence IDs must be unique")
        return self

    model_config = {"extra": "forbid", "frozen": True}


class DocxRenderLayer(BaseModel):
    layer_id: str
    target: DocxTextLocator
    replacement_text: str

    model_config = {"extra": "forbid", "frozen": True}


class DocxImageReplacement(BaseModel):
    occurrence_id: str = Field(pattern=r"^pic_[0-9a-f]{64}$")
    replacement_path: str = Field(min_length=1)
    replacement_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")

    model_config = {"extra": "forbid", "frozen": True}


class DocxRenderResult(BaseModel):
    document_sha256: str
    sanitation: SanitationSummary

    model_config = {"extra": "forbid", "frozen": True}

File: docx-processor/Protocol/Contracts.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Aiadr.Docx.Protocol;

internal sealed record RequestEnvelope(
    [property: JsonPropertyName("request_id")] string RequestId,
    [property: JsonPropertyName("operation")] string Operation,
    [property: JsonPropertyName("payload")] JsonElement Payload);

internal sealed record ErrorPayload(
    [property: JsonPropertyName("code")] string Code,
    [property: JsonPropertyName("message")] string Message,
    [property: JsonPropertyName("feature")] string? Feature = null);

internal sealed record ResponseEnvelope(
    [property: JsonPropertyName("request_id")] string RequestId,
    [property: JsonPropertyName("ok")] bool Ok,
    [property: JsonPropertyName("payload")] object? Payload,
    [property: JsonPropertyName("error")] ErrorPayload? Error,
    [property: JsonPropertyName("working_set_bytes")] long WorkingSetBytes);

internal sealed record HandshakeResponse(
    [property: JsonPropertyName("openxml_sdk_version")] string OpenXmlSdkVersion,
    [property: JsonPropertyName("operations")] string[] Operations);

internal sealed record InspectRequest(
    [property: JsonPropertyName("source_path")] string SourcePath,
    [property: JsonPropertyName("expected_source_sha256")] string ExpectedSourceSha256,
    [property: JsonPropertyName("output_path")] string OutputPath,
    [property: JsonPropertyName("image_output_directory")] string ImageOutputDirectory,
    [property: JsonPropertyName("allowed_roots")] string[] AllowedRoots);

internal sealed record RenderRequest(
    [property: JsonPropertyName("source_path")] string SourcePath,
    [property: JsonPropertyName("expected_source_sha256")] string ExpectedSourceSha256,
    [property: JsonPropertyName("output_path")] string OutputPath,
    [property: JsonPropertyName("layers")] DocxRenderLayer[] Layers,
    [property: JsonPropertyName("image_replacements")] DocxImageReplacement[] ImageReplacements,
    [property: JsonPropertyName("allowed_roots")] string[] AllowedRoots);

internal sealed record DocxRenderLayer(
    [property: JsonPropertyName("layer_id")] string LayerId,
    [property: JsonPropertyName("target")] DocxTextLocator Target,
    [property: JsonPropertyName("replacement_text")] string ReplacementText);

internal sealed record DocxTextLocator(
    [property: JsonPropertyName("format")] string Format,
    [property: JsonPropertyName("page")] int Page,
    [property: JsonPropertyName("line_id")] string LineId,
    [property: JsonPropertyName("source_sha256")] string SourceSha256,
    [property: JsonPropertyName("story_kind")] string StoryKind,
    [property: JsonPropertyName("part_uri")] string PartUri,
    [property: JsonPropertyName("block_id")] string BlockId,
    [property: JsonPropertyName("start")] int Start,
    [property: JsonPropertyName("end")] int End,
    [property: JsonPropertyName("exact_text")] string ExactText);

internal sealed record DocxTextBlock(
    [property: JsonPropertyName("block_id")] string BlockId,
    [property: JsonPropertyName("ordinal")] int Ordinal,
    [property: JsonPropertyName("story_kind")] string StoryKind,
    [property: JsonPropertyName("part_uri")] string PartUri,
    [property: JsonPropertyName("structural_path")] string StructuralPath,
    [property: JsonPropertyName("text")] string Text);

internal sealed record DocxImageOccurrence(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("ordinal")] int Ordinal,
    [property: JsonPropertyName("story_kind")] string StoryKind,
    [property: JsonPropertyName("part_uri")] string PartUri,
    [property: JsonPropertyName("media_type")] string MediaType,
    [property: JsonPropertyName("source_asset")] string? SourceAsset,
    [property: JsonPropertyName("crop_left")] int CropLeft,
    [property: JsonPropertyName("crop_top")] int CropTop,
    [property: JsonPropertyName("crop_right")] int CropRight,
    [property: JsonPropertyName("crop_bottom")] int CropBottom,
    [property: JsonPropertyName("flip_horizontal")] bool FlipHorizontal,
    [property: JsonPropertyName("flip_vertical")] bool FlipVertical,
    [property: JsonPropertyName("targetable")] bool Targetable,
    [property: JsonPropertyName("unsupported_reason")] string? UnsupportedReason);

internal sealed record DocxImageReplacement(
    [property: JsonPropertyName("occurrence_id")] string OccurrenceId,
    [property: JsonPropertyName("replacement_path")] string ReplacementPath,
    [property: JsonPropertyName("replacement_sha256")] string ReplacementSha256);

internal sealed record SanitationSummary(
    [property: JsonPropertyName("comments_removed")] int CommentsRemoved,
    [property: JsonPropertyName("properties_removed")] int PropertiesRemoved);

internal sealed record InspectResponse(
    [property: JsonPropertyName("character_count")] int CharacterCount,
    [property: JsonPropertyName("blocks")] DocxTextBlock[] Blocks,
    [property: JsonPropertyName("image_occurrences")] DocxImageOccurrence[] ImageOccurrences,
    [property: JsonPropertyName("document_sha256")] string DocumentSha256,
    [property: JsonPropertyName("sanitation")] SanitationSummary Sanitation);

internal sealed record RenderResponse(
    [property: JsonPropertyName("document_sha256")] string DocumentSha256,
    [property: JsonPropertyName("sanitation")] SanitationSummary Sanitation);

internal sealed class ProcessorException(string code, string message, string? feature = null)
    : Exception(message)
{
    public string Code { get; } = code;
    public string? Feature { get; } = feature;
}
